### ISSUE TITLE
conv_transpose1d weight constants deduplicated by shape — 2nd op silently reuses 1st's weights

### ISSUE THREAD

**Issue by john-rocky** [google-ai-edge/litert-torch#1061]
### Summary

Two `F.conv_transpose1d` ops whose weight **constants** share the same **shape** but have
**different values** are deduplicated during conversion: the second op silently reuses the
first op's weights. The trigger is weight-**value** dependent — the inverse-DFT cos/sin basis
pair (a normal inverse-STFT weight pair) gets merged, while random same-shape weights do not.

I hit this porting Kokoro-82M (TTS, ISTFTNet vocoder): it corrupts an in-graph inverse-STFT.
It reproduces on the default delegate **and** the reference kernels
(`BUILTIN_WITHOUT_DEFAULT_DELEGATES`), so it is conversion-level, not a runtime/XNNPACK issue.

### Environment

`torch 2.12.0` · `litert_torch 0.9.1` · `ai_edge_litert 2.1.5` · Python 3.11 · macOS arm64.

### Minimal repro

```python
import os
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import litert_torch
from ai_edge_litert.interpreter import Interpreter, OpResolverType


def corr(a, b):
    a, b = np.asarray(a).ravel(), np.asarray(b).ravel()
    n = min(len(a), len(b))
    return float(np.corrcoef(a[:n], b[:n])[0, 1])


def dft_inverse_bases(n_fft=20):
    """Inverse-DFT cos/sin bases (Hann window), each [n_fft//2+1, 1, n_fft]."""
    fb = n_fft // 2 + 1
    win = torch.hann_window(n_fft, periodic=True, dtype=torch.float32).numpy()
    n, k = np.arange(n_fft), np.arange(fb)
    ang = 2 * np.pi * np.outer(n, k) / n_fft
    inv_win = win * (1.0 / n_fft)
    W_cos = torch.from_numpy((np.cos(ang).T * inv_win)).float().unsqueeze(1)
    W_sin = torch.from_numpy((np.sin(ang).T * inv_win)).float().unsqueeze(1)
    return W_cos, W_sin


class TwoConvT(nn.Module):
    """out = conv_transpose1d(a, W1) - conv_transpose1d(b, W2)"""
    def __init__(self, W1, W2, stride):
        super().__init__()
        self.register_buffer("W1", W1)
        self.register_buffer("W2", W2)
        self.stride = stride

    def forward(self, a, b):
        return (F.conv_transpose1d(a, self.W1, stride=self.stride)
                - F.conv_transpose1d(b, self.W2, stride=self.stride)).squeeze()


def convert_and_run(m, args, rt):
    path = "/tmp/convT_dedup.tflite"
    if os.path.exists(path):
        os.remove(path)
    litert_torch.convert(m, tuple(args)).export(path)
    it = Interpreter(model_path=path, experimental_op_resolver_type=rt)
    it.allocate_tensors()
    for d, x in zip(it.get_input_details(), args):
        it.set_tensor(d["index"], x.numpy().astype(d["dtype"]))
    it.invoke()
    return it.get_tensor(it.get_output_details()[0]["index"])


def check(name, W1, W2, stride, in_ch, length):
    torch.manual_seed(0)
    a, b = torch.randn(1, in_ch, length), torch.randn(1, in_ch, length)
    m = TwoConvT(W1, W2, stride).eval()
    with torch.no_grad():
        correct = m(a, b).numpy()
        wrong = (F.conv_transpose1d(a, W1, stride=stride)
                 - F.conv_transpose1d(b, W1, stride=stride)).squeeze().numpy()  # 2nd reuses W1
    print(f"\n== {name}  weights={tuple(W1.shape)} stride={stride} ==")
    for rt in (OpResolverType.AUTO, OpResolverType.BUILTIN_WITHOUT_DEFAULT_DELEGATES):
        tfl = convert_and_run(m, (a, b), rt)
        c_ok, c_bug = corr(tfl, correct), corr(tfl, wrong)
        verdict = "BUG (2nd conv reuses 1st weights)" if c_bug > 0.999 and c_ok < 0.99 else "ok"
        print(f"   {str(rt).split('.')[-1]:34s} corr_vs_correct={c_ok:.4f}  corr_vs_reused-W1={c_bug:.4f}  -> {verdict}")


W_cos, W_sin = dft_inverse_bases(20)
check("CASE A (DFT cos/sin bases)", W_cos, W_sin, stride=5, in_ch=11, length=200)
torch.manual_seed(1)
check("CASE B control (random, same shape)", torch.randn(11, 1, 20), torch.randn(11, 1, 20),
      stride=5, in_ch=11, length=200)
```

### Observed

```
== CASE A (DFT cos/sin bases)  weights=(11, 1, 20) stride=5 ==
   AUTO                               corr_vs_correct=0.5587  corr_vs_reused-W1=1.0000  -> BUG (2nd conv reuses 1st weights)
   BUILTIN_WITHOUT_DEFAULT_DELEGATES  corr_vs_correct=0.5587  corr_vs_reused-W1=1.0000  -> BUG (2nd conv reuses 1st weights)
== CASE B control (random, same shape)  weights=(11, 1, 20) stride=5 ==
   AUTO                               corr_vs_correct=1.0000  corr_vs_reused-W1=0.5077  -> ok
   BUILTIN_WITHOUT_DEFAULT_DELEGATES  corr_vs_correct=1.0000  corr_vs_reused-W1=0.5077  -> ok
```

The converted CASE-A output matches "both convs use W_cos" at corr **1.0000** (vs 0.56 to the
correct `cos − sin`). CASE B (random, identical shapes) converts correctly — so a constant-dedup
pass is incorrectly collapsing the two distinct (DFT-structured) weight constants.

### Impact

Corrupts any in-graph inverse-STFT / vocoder / dual-branch conv built from a shared-shape
weight pair (very common in TTS/audio). For Kokoro-82M I worked around it by moving the final
iSTFT host-side (numerically exact, ~free), but the in-graph path is silently wrong.

### Expected

Both cases convert to `corr_vs_correct ~1.0`; weight constants must not be merged by value/shape.

**Comment by john-rocky** [comment]
Re-verified on `main` HEAD (commit `adfe360`): this still reproduces. CASE A (the cos/sin inverse-DFT bases) converts to `corr_vs_correct=0.5587 / corr_vs_reused-W1=1.0000` on both the default delegate and `BUILTIN_WITHOUT_DEFAULT_DELEGATES`.

One correction to the title: it isn't deduplication *by shape* — random same-shape weights are **not** merged. The root cause is in `litert_torch/backend/inline_consts.py`. The constant cache is keyed by `_tensor_fingerprint`, which uses `tensor.untyped_storage().data_ptr()`, computed on the temporary copy from `x = x.contiguous().detach().cpu()`:

```python
def _tensor_fingerprint(tensor):
    return (str(tensor.device), tensor.shape, tensor.stride(),
            tensor.untyped_storage().data_ptr())
```

That temporary copy is freed when the lowering returns, and the allocator hands the **same address** to the next constant — so two distinct constants collide on the cache key and the second silently reuses the first's MLIR attribute.

Instrumented proof (cos/sin case):

| tensor | original storage ptr | temp-copy ptr (the cache key) | sum |
|---|---|---|---|
| W_cos | ...630400 | **...135936** | 0.2500 |
| W_sin | ...631296 | **...135936** ← collides | 0.0000 |

It only looks "value-dependent" because address reuse is allocation-timing dependent (random pairs happened not to collide). The merge happens inside `exported_program_to_mlir`, before the converter — both StableHLO constants end up carrying W_cos's bytes (`attr0 == attr1` is `True`).

Fix: key the cache on tensor **content** instead of a storage address. I have a patch plus a regression test ready and will open a PR.

**Comment by shivvaam0001** [comment]
Hi @john-rocky 

Thank you for this detailed bug report and for identifying the root cause. We will investigate the issue and get back to you.

Thanks!

**Comment by shivvaam0001** [comment]
Hi @john-rocky 

Thanks for flagging the issue. I am able to successfully reproduce it and observed the same following logs:
```

(00:00) [ DONE] Write Model to /tmp/convT_dedup.tflite (+00:00)
   AUTO                               corr_vs_correct=0.5587  corr_vs_reused-W1=1.0000  -> BUG (2nd conv reuses 1st weights)
(00:00) [ DONE] Write Model to /tmp/convT_dedup.tflite (+00:00)
   BUILTIN_WITHOUT_DEFAULT_DELEGATES  corr_vs_correct=0.5587  corr_vs_reused-W1=1.0000  -> BUG (2nd conv reuses 1st weights)

```

I also tried applying the patch you proposed. And here what i observed:

```
(00:00) [ DONE] Write Model to /tmp/convT_dedup.tflite (+00:00)
   AUTO                               corr_vs_correct=1.0000  corr_vs_reused-W1=0.5587  -> ok
00:00) [ DONE] Write Model to /tmp/convT_dedup.tflite (+00:00)
   BUILTIN_WITHOUT_DEFAULT_DELEGATES  corr_vs_correct=1.0000  corr_vs_reused-W1=0.5587  -> ok

```

**Comment by john-rocky** [comment]
Thanks for reproducing it and for testing the patch, @shivvaam0001 — your post-patch numbers are exactly the expected signature: `corr_vs_correct=1.0000` with `corr_vs_reused-W1` dropping to 0.5587 means the two `conv_transpose1d` ops stopped sharing a constant.

The fix is already up as **#1063** (Fixes #1061), open since 2026-06-19 with no reviewer assigned yet.

Root cause, for the record: `inline_consts._tensor_fingerprint` keyed the constant cache on `tensor.untyped_storage().data_ptr()`, computed on the temporary produced by `x.contiguous().detach().cpu()`. That temporary is freed when the lowering returns, and the allocator hands the same address to the next constant → false cache hit → two distinct constants merge into one MLIR attribute. That's also why it looked weight-*value* dependent: in the random control case both tensors stay alive, so no address gets recycled and nothing merges.

The PR keys the cache on tensor **content** instead (blake2b-128 over the raw bytes via `view(torch.uint8)`, so it's dtype-agnostic incl. bfloat16), which also makes the intended cross-signature dedup more correct — genuinely identical constants now share even across distinct tensor objects. It adds `litert_torch/backend/test/test_inline_consts.py` with 3 regression tests, including one that fails on the old key (same storage, different content).

Could someone assign a reviewer? Happy to rebase if it's gone stale in the meantime.

**Comment by aravindmurali22** [comment]
Hi @markoristicc , following up on this thread, is there any update on this issue? Thanks!

**Comment by john-rocky** [comment]
Thanks for reproducing this and for testing the patch, @shivvaam0001. Two more models I am porting hit the same collision, so I am adding both repros here rather than opening new issues. The new part is the trigger: any **non-contiguous** constant takes this path.

**Why non-contiguous.** The lowering calls `x.contiguous()` before it fingerprints a constant, and the fingerprint still includes the storage address. A contiguous constant keeps its own storage, so its address is stable. A non-contiguous one gets a temporary, and two temporaries of the same shape and stride land on the same address one after the other. That is the address reuse described above. Three ways a model gets there:

- a weight built as `torch.Tensor(np_array.T)`: torchlibrosa keeps the transposed strides;
- `.real` / `.imag` views of one complex buffer: the SAM 3 ViT stores its RoPE cos and sin this way;
- `.t()` views.

The contiguous form of each converts correctly. In every run with a hook on `_ConstantFingerprint.from_tensor`, the logged false cache hits are on exactly the pairs below and on nothing else.

**The two models** (fp32, CPU, default delegate and reference kernels give the same numbers; PyTorch is the reference):

| model | the non-contiguous pair | what the converted graph does | corr vs PyTorch | with #1063 |
|---|---|---|---|---|
| PANNs CNN14: `torchlibrosa.stft.Spectrogram` alone (n_fft 1024, hop 320, 10 s of audio) | the cos and sin DFT bases of two `Conv1d`, built as `torch.Tensor(np.T)` | one `DEPTHWISE_CONV_2D` where two belong; the output matches "cos basis in both convs" at corr 1.0000 | 0.71 (random audio), 0.83 (speech) | 1.0000, two convs again |
| SAM 3 ViT-L/14: stock `vitdet` blocks and the full vision graph, input [1, 72, 72, 1024] | RoPE cos and sin as `.real` / `.imag` of one `freqs_cis` buffer | RoPE rotates with cos in both slots | window block 0.9949, global block 0.8056; full vision graph 0.51 to 0.56 on a real photo, with 31 or 32 of the 32 RoPE pairs collapsed onto one buffer | 1.0000 for both blocks and the full graph |

**Environment.** `litert_torch 0.9.4` (PyPI; torch 2.13.0, Python 3.14) and `main` at d592a2f (0.10.0; torch 2.14.0, Python 3.12), `ai_edge_litert 2.2.0`, macOS arm64. The PANNs numbers and the script below are the same on both; the SAM 3 numbers are from main, and the full graph also reproduces on 0.9.3.

**Minimal repro** (the two patterns, both resolvers; `corr_vs_aliased` compares with the second constant replaced by the first):

```python
import numpy as np, torch, torch.nn as nn, litert_torch
from ai_edge_litert.interpreter import Interpreter, OpResolverType


class ComplexViews(nn.Module):
    """sam3 vitdet form: cos and sin as .real / .imag views of one complex buffer."""

    def __init__(self, contiguous):
        super().__init__()
        c = torch.polar(torch.ones(64, 8), torch.rand(64, 8) * 6.2832)
        self.register_buffer("cos", c.real.clone() if contiguous else c.real)
        self.register_buffer("sin", c.imag.clone() if contiguous else c.imag)

    def forward(self, x, aliased=False):
        second = self.cos if aliased else self.sin
        return x * self.cos + 2.0 * (x * second)


class TransposedParams(nn.Module):
    """torchlibrosa form: torch.Tensor(np_array.T) keeps the transposed strides."""

    def __init__(self, contiguous):
        super().__init__()
        a, b = np.random.default_rng(0).standard_normal((2, 8, 64))
        ta, tb = torch.Tensor(a.T), torch.Tensor(b.T)
        self.a = nn.Parameter(ta.contiguous() if contiguous else ta, requires_grad=False)
        self.b = nn.Parameter(tb.contiguous() if contiguous else tb, requires_grad=False)

    def forward(self, x, aliased=False):
        second = self.a if aliased else self.b
        return x @ self.a + 2.0 * (x @ second)


def corr(a, b):
    return float(np.corrcoef(np.ravel(a), np.ravel(b))[0, 1])


def check(name, m, x, contiguous):
    with torch.no_grad():
        correct, aliased = m(x).numpy(), m(x, aliased=True).numpy()
    litert_torch.convert(m.eval(), (x,)).export("/tmp/alias.tflite")
    for rt in (OpResolverType.AUTO, OpResolverType.BUILTIN_WITHOUT_DEFAULT_DELEGATES):
        it = Interpreter(model_path="/tmp/alias.tflite", experimental_op_resolver_type=rt)
        it.allocate_tensors()
        it.set_tensor(it.get_input_details()[0]["index"], x.numpy())
        it.invoke()
        y = it.get_tensor(it.get_output_details()[0]["index"])
        c_ok, c_bug = corr(y, correct), corr(y, aliased)
        verdict = "BUG (2nd constant reuses 1st)" if c_bug > 0.999 and c_ok < 0.99 else "ok"
        print(f"{name:17s} contiguous={contiguous!s:5s} {str(rt).split('.')[-1]:34s} "
              f"corr_vs_correct={c_ok:.4f}  corr_vs_aliased={c_bug:.4f}  -> {verdict}")


torch.manual_seed(0)
x1, x2 = torch.randn(1, 64, 8), torch.randn(1, 4, 64)
for contiguous in (False, True):
    check("ComplexViews", ComplexViews(contiguous), x1, contiguous)
    check("TransposedParams", TransposedParams(contiguous), x2, contiguous)
```

Observed:

```
ComplexViews      contiguous=False AUTO                               corr_vs_correct=0.3554  corr_vs_aliased=1.0000  -> BUG (2nd constant reuses 1st)
ComplexViews      contiguous=False BUILTIN_WITHOUT_DEFAULT_DELEGATES  corr_vs_correct=0.3554  corr_vs_aliased=1.0000  -> BUG (2nd constant reuses 1st)
TransposedParams  contiguous=False AUTO                               corr_vs_correct=0.3373  corr_vs_aliased=1.0000  -> BUG (2nd constant reuses 1st)
TransposedParams  contiguous=False BUILTIN_WITHOUT_DEFAULT_DELEGATES  corr_vs_correct=0.3373  corr_vs_aliased=1.0000  -> BUG (2nd constant reuses 1st)
ComplexViews      contiguous=True  AUTO                               corr_vs_correct=1.0000  corr_vs_aliased=0.4101  -> ok
ComplexViews      contiguous=True  BUILTIN_WITHOUT_DEFAULT_DELEGATES  corr_vs_correct=1.0000  corr_vs_aliased=0.4101  -> ok
TransposedParams  contiguous=True  AUTO                               corr_vs_correct=1.0000  corr_vs_aliased=0.3373  -> ok
TransposedParams  contiguous=True  BUILTIN_WITHOUT_DEFAULT_DELEGATES  corr_vs_correct=1.0000  corr_vs_aliased=0.3373  -> ok
```

With #1063 applied (its diff patched into both installs, then reverted), every row above prints `ok` at corr 1.0000, and the hook logs zero false hits for either model. The PR has no conflicts against today's main. If a different shape suits the tree better, tell me and I will reshape it.

Thanks again for the quick reproduction.
